### PR TITLE
Add type annotations for ArrayPatterns and ObjectPatterns.

### PR DESCRIPTION
**Pull request by gabelevi**
See the test for examples.

Should I initialize the typeAnnotation to `null` or `undefined` for all ObjectPatterns and ArrayPatterns? I didn't initially since it was a pain to update all the test cases (is there any good way to re-record a whole test fixture? All I found was  tools/generate-test-fixture.js

### CODE REVIEW

**Comment by gabelevi** [comment]
Ping @jeffmo

File: test/fbtest.js [hunk 1]
```diff
@@ -5859,7 +5859,440 @@ var fbTestFixture = {
                 start: { line: 1, column: 0 },
                 end: { line: 1, column: 41 }
             }
-        }
+        },
+        'var {x}: {x: string; } = { x: "hello" };': {
```
Comment (jeffmo) [inline test/fbtest.js]: Is it a syntax error to leave out the semicolon in a type expression that has only one key/val entry? (either way, can we add a test to that effect?)

File: test/fbtest.js [hunk 2]
```diff
@@ -5859,7 +5859,440 @@ var fbTestFixture = {
...
+              }
+          }],
+          kind: 'var',
+          range: [0, 40],
+          loc: {
+              start: { line: 1, column: 0 },
+              end: { line: 1, column: 40 }
+          }
+        },
+        'var [x]: Array<string> = [ "hello" ];': {
```
Comment (jeffmo) [inline test/fbtest.js:6652]: Support for `var [x]: string[] = [ "hello" ];` too, right?

File: test/fbtest.js (hunk 1 above)
Comment (gabelevi) [inline test/fbtest.js]: Sure, that's easy enough...one moment...

File: test/fbtest.js (hunk 2 above)
Comment (gabelevi) [inline test/fbtest.js:6652]: Yeah, that's the plan eventually, but it's not in yet.

**Comment by gabelevi** [comment]
This passes the esprima tester, btw

File: test/fbtest.js [hunk 3]
```diff
@@ -5940,3 +6516,6 @@ var fbTestFixture = {
     }
 
 }());
+
```
Comment (jeffmo) [inline test/fbtest.js]: such newlines. much whitespace.

**Comment by gabelevi** [comment]
Also I rebased on top of pull requests that were recently merged

**Pull request merged**